apps/voting/src/shared/store.ts

```diff
@@ -119,48 +122,115 @@ const electionConfigBaseAtom = atom<ElectionConfig | undefined>(undefined);
+            };
         }
 
-        const electionManifestPromise = getChecksumResource<ElectionManifest>(config.election_manifest);
-        const electionParametersPromise = getChecksumResource<ElectionParameters>(config.election_parameters);
-        const candidatePromises = config.candidates.map(getCandidate);
-
+        const electionManifestPromise = getChecksumResource<ElectionManifest>(preliminaryConfig.manifest);
+        const electionParametersPromise = getChecksumResource<ElectionParameters>(preliminaryConfig.parameters);
+        const candidatePromises = preliminaryConfig.candidates.map(getCandidate);
         const [manifest, parameters, ...candidates] = await Promise.all([
             electionManifestPromise,
             electionParametersPromise,
             ...candidatePromises,
         ]);
 
-        const mappedConfig: ElectionConfig = {
-            start: Timestamp.toDate(config.election_start),
-            end: Timestamp.toDate(config.election_end),
+        const config: ElectionConfig = {
+            ...preliminaryConfig,
             candidates: candidates.filter(isDefined),
-            description: config.election_description,
-            manifest,
-            parameters,
-            guardianKeys: config.guardian_keys,
-            setupDone: false,
+            manifest: manifest,
+            parameters: parameters,
         };
+        set(electionConfigBaseAtom, config);
+    },
+);
 
-        set(electionConfigBaseAtom, mappedConfig);
+/**
+ * Primitive atom for holding the {@linkcode ElectionConfig} of the election contract
+ */
+const guardiansStateBaseAtom = atom<GuardiansState | undefined>(undefined);
 
-        const guardians = await getGuardiansState();
-        if (guardians === undefined) {
-            throw new Error('Failed to get election config');
+/**
+ * Holds the configuration of the election contract. A reference to this should always be kept in the application root
+ * to avoid having to fetch the configuration more than once.
+ */
+export const guardiansStateAtom = atom(
+    (get) => get(guardiansStateBaseAtom),
+    async (_, set) => {
+        const hydrate = CONFIG.contractConfig !== undefined;
+        let config: GuardiansState;
+        if (hydrate) {
+            config = {
+                guardianKeys: CONFIG.contractConfig!.guardian_keys,
+                setupDone: CONFIG.contractConfig!.guardians_setup_done,
+            };
+        } else {
+            const guardians = await getGuardiansState();
+            if (guardians === undefined) {
+                throw new Error('Failed to get election config');
+            }
+
+            const setupDone = guardians.every(
+                ([, g]) =>
+                    g.public_key.type === 'Some' &&
+                    g.encrypted_share.type === 'Some' &&
+                    g.status.type === 'Some' &&
+                    g.status.content.type === 'VerificationSuccessful',
+            );
```

Comment:
Is the backend also checking/calculating if the guardian setup is done? We can use that value instead of recomputing it, isn't it?

Reply:
Yes so we do also use the calculation from the backend if it is present. The `else` branch of the `if (hydrate)` statement only runs if the configuration passed from the backend is _not_ there, i.e. if we run the application with the development server. Hydrate in this case means to fill the frontend state with state transferred from the backend (it's a real term 😄)

This is a recurring approach in `electionConfigAtom`, `guardiansStateAtom` and `electionResultAtom`; they all try to use state constructed in the backend if available.